# [M2S2] Ex. 01 - Chipotle Dataset

**Você é o Engenheiro de Dados da rede Chipotle. O sistema de vendas sofreu uma pane e gerou logs "sujos". O CEO, o Gerente de Estoque e o Diretor de Marketing precisam de respostas urgentes, mas o Excel não consegue processar os dados.**

In [19]:
# 0 - Carregar os dados (chipotle_modificado.csv)

## Descrição: Chipotle Orders
## Contexto: Este dataset representa um log de transações de vendas da rede de fast-food mexicana Chipotle.
## Cada linha representa um item comprado, não necessariamente um pedido inteiro.
## Imagine que esta tabela é o espelho digital da "Nota Fiscal" que sai da impressora.

# Detalhes da tabela:
## COLUNA                DESCRIÇAO
## order_id              O número do cupom fiscal. Se você comprou 3 coisas, esse número se repete em 3 linhas.
## quantity	             Quantas unidades daquele item específico foram pedidas.
## item_name	           O nome principal do produto.
## choice_description	   Os detalhes/customização do pedido (molhos, acompanhamentos).
## item_price	           O valor cobrado pelo item.

In [20]:
import pandas as pd
df_logs = pd.read_csv('../../data/input/csv/chipotle_modificado.csv')
print(f"Linhas totais importadas: {df_logs.shape[0]}")
print(f"Colunas detectadas: {df_logs.columns.tolist()}")
print("--- Visualizando as primeiras linhas ---")
display(df_logs.head())

Linhas totais importadas: 4622
Colunas detectadas: ['order_id', 'quantity', 'item_name', 'choice_description', 'item_price']
--- Visualizando as primeiras linhas ---


,order_id,quantity,item_name,choice_description,item_price
0,1,1.0,Chips and Fresh Tomato Salsa,NaN,$2.39
1,1,1.0,Izze,[Clementine],$3.39
2,1,1.0,Nantucket Nectar,[Apple],$3.39
3,1,1.0,Chips and Tomatillo-Green Chili Salsa,NaN,$2.39
4,2,2.0,Chicken Bowl,"[Tomatillo-Red Chili Salsa (Hot), [Black Beans...",$16.98


In [21]:
df_logs.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4622 entries, 0 to 4621
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   order_id            4622 non-null   int64  
 1   quantity            4395 non-null   float64
 2   item_name           4622 non-null   object 
 3   choice_description  3376 non-null   object 
 4   item_price          4622 non-null   object 
dtypes: float64(1), int64(1), object(3)
memory usage: 180.7+ KB


In [22]:
df_logs.tail()

,order_id,quantity,item_name,choice_description,item_price
4617,1833,1.0,Steak Burrito,"[Fresh Tomato Salsa, [Rice, Black Beans, Sour ...",$11.75
4618,1833,1.0,Steak Burrito,"[Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...",$11.75
4619,1834,1.0,Chicken Salad Bowl,"[Fresh Tomato Salsa, [Fajita Vegetables, Pinto...",$11.25
4620,1834,1.0,Chicken Salad Bowl,"[Fresh Tomato Salsa, [Fajita Vegetables, Lettu...",$8.75
4621,1834,1.0,Chicken Salad Bowl,"[Fresh Tomato Salsa, [Fajita Vegetables, Pinto...",$8.75


In [23]:
df_logs.sample(5)

,order_id,quantity,item_name,choice_description,item_price
1271,517,1.0,Chicken Burrito,"[Tomatillo Red Chili (Hot), [Lettuce, Rice, Ch...",$8.19
2878,1143,1.0,Canned Soft Drink,[Coke],$1.25
1944,787,2.0,Canned Soda,[Dr. Pepper],$2.18
859,354,1.0,Barbacoa Burrito,"[Fresh Tomato Salsa, [Fajita Vegetables, Rice,...",$9.25
2202,888,1.0,Canned Soft Drink,[Sprite],$1.25


In [24]:
df_logs.describe()

,order_id,quantity
count,4622.000000,4395.000000
mean,927.254868,1.077361
std,528.890796,0.417989
min,1.000000,1.000000
25%,477.250000,1.000000
50%,926.000000,1.000000
75%,1393.000000,1.000000
max,1834.000000,15.000000


Atividade 1:
**O Mistério do Faturamento**

(Limpeza de Dados e Tipos)  |   Stakeholder: CEO / Financeiro.

O CEO quer saber: "Quanto vendemos no total?"

In [25]:
print("Revisando o tipo original da coluna item_price")
print(df_logs['item_price'].dtype)
print(df_logs['item_price'].head(10))

Revisando o tipo original da coluna item_price
object
0     $2.39 
1     $3.39 
2     $3.39 
3     $2.39 
4    $16.98 
5    $10.98 
6     $1.69 
7    $11.75 
8     $9.25 
9     $9.25 
Name: item_price, dtype: object


Para aplicar transformações e manter o dataset original, fazemos uma cópia.

In [26]:
df_logs_transform = df_logs.copy()
df_logs_transform.head()

,order_id,quantity,item_name,choice_description,item_price
0,1,1.0,Chips and Fresh Tomato Salsa,NaN,$2.39
1,1,1.0,Izze,[Clementine],$3.39
2,1,1.0,Nantucket Nectar,[Apple],$3.39
3,1,1.0,Chips and Tomatillo-Green Chili Salsa,NaN,$2.39
4,2,2.0,Chicken Bowl,"[Tomatillo-Red Chili Salsa (Hot), [Black Beans...",$16.98


Verificamos o tipo errado para cáclulos da coluna "item_price" (como object ou texto). Portanto, precisamos realizar uma transformação de tipo, retirando o "$" dos valores.

In [27]:
df_logs_transform['item_price'] = df_logs_transform['item_price'].astype(str).str.replace('$', '').str.replace(',', '.').astype(float)
df_logs_transform.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4622 entries, 0 to 4621
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   order_id            4622 non-null   int64  
 1   quantity            4395 non-null   float64
 2   item_name           4622 non-null   object 
 3   choice_description  3376 non-null   object 
 4   item_price          4622 non-null   float64
dtypes: float64(2), int64(1), object(2)
memory usage: 180.7+ KB


Uma vez transformado, é possível agora realizar o cálculo do valor total de cada item numa nova coluna. 

In [28]:
df_logs_transform['valor_item'] = df_logs_transform['quantity'] * df_logs_transform['item_price']

Nessa nova coluna guardamos os valores ideais para o cálculo de faturamento total.

In [38]:
faturamento_total = df_logs_transform['valor_item'].sum()
print("="*50)
print(f"Faturamento Total: $ {faturamento_total:,.2f}")
print(f"Total de Itens Vendidos: {df_logs_transform['quantity'].sum()}")
print(f"Ticket Médio por Item: $ {faturamento_total / df_logs_transform.shape[0]:,.2f}")
print("="*50)

Faturamento Total: $ 37,362.02
Total de Itens Vendidos: 4962.0
Ticket Médio por Item: $ 8.08


Atividade 2: **Auditoria de Estoque**
(Tratamento de NaNs e Estatística)

Stakeholder: Gerente de Estoque.

Precisamos repor o estoque respondendo a pergunta:

"A cada 100 vendas realizadas, quantas unidades deste produto saíram da prateleira?"


In [30]:
print("Verificando valores ausentes em quantity:")
print(f"Valores NaN: {df_logs_transform['quantity'].isna().sum()}")
print(f"Total de registros: {df_logs_transform.shape[0]}")

Verificando valores ausentes em quantity:
Valores NaN: 227
Total de registros: 4622


Percebemos na coluna "quantity" a presença de 227 valores faltantes (NaN). Isso indica a necessidade de tratamento antes de realizar cálculos estatísticos. As principais estratégias disponíveis são:

* **Remover os NaN**: Eliminar as linhas com valores ausentes (abordagem conservadora)
* **Preencher com 1**: Assumir que registros sem quantidade = 1 unidade vendida
* **Preencher com mediana**: Substituir por um valor estatístico representativo

Dentro do contexto deste dataset, não há informações suficientes para determinar a causa exata desses valores ausentes. Entretanto, como cada linha possui um `item_price` registrado (indicando que houve uma transação real), a estratégia mais pragmática é preencher com **1 unidade**. 

Afinal, se existe um registro de venda no sistema com valor monetário, é razoável inferir que ao menos 1 item foi vendido, mesmo que a quantidade não tenha sido capturada corretamente pelo sistema.

In [31]:
print(f"\nDecisão: Preenchendo {df_logs_transform['quantity'].isna().sum()} valores NaN com 1")
df_logs_transform['quantity'].fillna(1, inplace=True)

print(f"Valores NaN restantes: {df_logs_transform['quantity'].isna().sum()}")


Decisão: Preenchendo 227 valores NaN com 1
Valores NaN restantes: 0


Com os dados limpos, podemos agora calcular a métrica solicitada pelo Gerente de Estoque: quantas unidades saem da prateleira a cada 100 vendas realizadas.

In [34]:
total_vendas = df_logs_transform.shape[0]
total_unidades = df_logs_transform['quantity'].sum()
unidades_por_100_vendas = (total_unidades / total_vendas) * 100

print("="*60)
print(f"Total de Vendas (registros): {total_vendas}")
print(f"Total de Unidades Vendidas: {int(total_unidades)}")
print(f"Média de Unidades por Venda: {total_unidades / total_vendas:.2f}")
print(f"\nA cada 100 vendas: {unidades_por_100_vendas:.2f} unidades saem da prateleira")
print("="*60)

Total de Vendas (registros): 4622
Total de Unidades Vendidas: 4962
Média de Unidades por Venda: 1.07

A cada 100 vendas: 107.36 unidades saem da prateleira


Para complementar a análise, vamos identificar quais produtos têm maior saída de estoque:

In [35]:
analise_estoque = df_logs_transform.groupby('item_name').agg({
    'quantity': ['sum', 'mean', 'count']
}).round(2)

analise_estoque.columns = ['Total_Unidades', 'Media_por_Venda', 'Num_Vendas']
analise_estoque = analise_estoque.sort_values('Total_Unidades', ascending=False)

print("\nTop 10 Produtos com Maior Saída de Estoque:")
print(analise_estoque.head(10))


Top 10 Produtos com Maior Saída de Estoque:
                              Total_Unidades  Media_por_Venda  Num_Vendas
item_name                                                                
Chicken Bowl                           759.0             1.05         726
Chicken Burrito                        588.0             1.06         553
Chips and Guacamole                    506.0             1.06         479
Steak Burrito                          384.0             1.04         368
Canned Soft Drink                      351.0             1.17         301
Chips                                  229.0             1.09         211
Steak Bowl                             221.0             1.05         211
Bottled Water                          211.0             1.30         162
Chips and Fresh Tomato Salsa           130.0             1.18         110
Canned Soda                            125.0             1.20         104


**Insights para o Gerente de Estoque:**

Analisando os 10 produtos com maior saída de estoque, podemos observar que:

1. **Chicken Bowl** lidera disparado com 759 unidades vendidas em 726 transações, representando o produto que mais demanda reposição.

2. **Padrão de consumo por categoria:**
   - **Pratos principais** (Bowls e Burritos): Chicken Bowl, Chicken Burrito, Steak Burrito e Steak Bowl concentram a maior parte das vendas, com médias próximas de 1.05 unidades por venda.
   - **Acompanhamentos**: Chips and Guacamole aparece em 3º lugar com 506 unidades, indicando alto consumo de complementos.
   - **Bebidas**: Canned Soft Drink (351 un.) e Bottled Water (211 un.) têm médias mais altas (1.17 e 1.30), sugerindo que clientes tendem a comprar múltiplas bebidas.

3. **Recomendação de reposição prioritária:**
   - Top 3 crítico: Chicken Bowl, Chicken Burrito e Chips and Guacamole devem ter estoque reforçado diariamente.
   - Bebidas enlatadas apresentam média de 1.17-1.30 unidades/venda, indicando compra em quantidade (provavelmente para grupos).

4. **Observação estratégica**: A baixa variação nas médias (entre 1.04 e 1.30) sugere que a maioria das vendas é de 1 unidade, com exceções para bebidas onde há compra de múltiplas unidades.

Atividade 3: O Clube VIP (Granularidade e Lógica)
(Agrupamento e Feature Engineering)
Stakeholder: Marketing / CRM.

A Dor do Negócio: Queremos lançar o programa "Chipotle Gold".

A regra é: Pedidos com valor total acima de R$20.00 são considerados VIP. Qual a % do nosso faturamento vem desses clientes VIP?

In [39]:
valor_por_pedido = df_logs_transform.groupby('order_id').agg({
    'valor_item': 'sum'
}).round(2)

valor_por_pedido.columns = ['valor_total_pedido']

print("Primeiros 10 pedidos e seus valores totais:")
print(valor_por_pedido.head(10))
print(f"\nTotal de Pedidos Únicos: {valor_por_pedido.shape[0]}")
print(f"Valor Mínimo de Pedido: $ {valor_por_pedido['valor_total_pedido'].min():.2f}")
print(f"Valor Máximo de Pedido: $ {valor_por_pedido['valor_total_pedido'].max():.2f}")
print(f"Valor Médio de Pedido: $ {valor_por_pedido['valor_total_pedido'].mean():.2f}")

Primeiros 10 pedidos e seus valores totais:
          valor_total_pedido
order_id                    
1                      11.56
2                      33.96
3                      12.67
4                      21.00
5                      13.70
6                      17.50
7                      15.70
8                      10.88
9                      12.85
10                     13.20

Total de Pedidos Únicos: 1834
Valor Mínimo de Pedido: $ 0.00
Valor Máximo de Pedido: $ 1074.24
Valor Médio de Pedido: $ 20.37


Observando os dados, vemos que:
- Temos 1834 pedidos únicos
- O valor médio é $ 20.37, muito próximo ao limiar de VIP ($20.00)
- Há uma grande variação: desde $ 0.00 até $ 1074.24

Agora vamos identificar quais pedidos são VIP (acima de $20.00) e calcular o impacto no faturamento.

In [40]:
valor_por_pedido['eh_vip'] = valor_por_pedido['valor_total_pedido'] > 20.00

total_pedidos = valor_por_pedido.shape[0]
pedidos_vip = valor_por_pedido['eh_vip'].sum()
pedidos_regular = total_pedidos - pedidos_vip

faturamento_vip = valor_por_pedido[valor_por_pedido['eh_vip']]['valor_total_pedido'].sum()
faturamento_regular = valor_por_pedido[~valor_por_pedido['eh_vip']]['valor_total_pedido'].sum()
faturamento_total_vip = faturamento_vip + faturamento_regular

percentual_vip = (faturamento_vip / faturamento_total_vip) * 100

print("="*70)
print("ANÁLISE: PROGRAMA CHIPOTLE GOLD (VIP)")
print("="*70)
print(f"\nTotal de Pedidos: {total_pedidos}")
print(f"  - Pedidos VIP (> $20.00): {int(pedidos_vip)} ({(pedidos_vip/total_pedidos)*100:.1f}%)")
print(f"  - Pedidos Regular (≤ $20.00): {int(pedidos_regular)} ({(pedidos_regular/total_pedidos)*100:.1f}%)")

print(f"\nFaturamento Total: $ {faturamento_total_vip:,.2f}")
print(f"  - Faturamento VIP: $ {faturamento_vip:,.2f}")
print(f"  - Faturamento Regular: $ {faturamento_regular:,.2f}")

print(f"\n{'RESPOSTA PARA O MARKETING':^70}")
print(f"Os clientes VIP representam {percentual_vip:.1f}% do faturamento total!")
print("="*70)

ANÁLISE: PROGRAMA CHIPOTLE GOLD (VIP)

Total de Pedidos: 1834
  - Pedidos VIP (> $20.00): 574 (31.3%)
  - Pedidos Regular (≤ $20.00): 1260 (68.7%)

Faturamento Total: $ 37,362.02
  - Faturamento VIP: $ 20,574.13
  - Faturamento Regular: $ 16,787.89

                      RESPOSTA PARA O MARKETING                       
Os clientes VIP representam 55.1% do faturamento total!


**Insights Estratégicos para o Programa Chipotle Gold:**

Os resultados revelam uma oportunidade significativa para o programa de fidelização:

1. **Proporção VIP vs Regular:**
   - Apenas 31.3% dos pedidos são VIP (574 de 1834)
   - Mas esses 31.3% geram **55.1% do faturamento total**
   - Isso demonstra que clientes VIP têm valor 1.76x superior aos clientes regulares

2. **Impacto Financeiro:**
   - Clientes VIP: $ 20,574.13
   - Clientes Regular: $ 16,787.89
   - Diferença: $ 3,786.24 (esse é o valor agregado pelos VIPs)

3. **Estratégia Recomendada:**
   - O programa Chipotle Gold deve focar em **converter clientes Regular em VIP**
   - Pequenos incentivos para atingir $20.00 podem gerar grande retorno
   - Os VIPs já existem e são altamente lucrativos - manter satisfação é crítico

4. **Oportunidade de Crescimento:**
   - Existem 1260 clientes no segmento Regular
   - Se apenas 10% deles migrar para VIP, adiciona ~$1,340 em receita

In [41]:
valor_medio_vip = valor_por_pedido[valor_por_pedido['eh_vip']]['valor_total_pedido'].mean()
valor_medio_regular = valor_por_pedido[~valor_por_pedido['eh_vip']]['valor_total_pedido'].mean()

quase_vip = valor_por_pedido[(valor_por_pedido['valor_total_pedido'] >= 18) & 
                             (valor_por_pedido['valor_total_pedido'] <= 20)]

print("\n" + "="*70)
print("ANÁLISE COMPLEMENTAR: OPORTUNIDADES DE CONVERSÃO")
print("="*70)

print(f"\nValor Médio por Segmento:")
print(f"  - VIP: $ {valor_medio_vip:.2f}")
print(f"  - Regular: $ {valor_medio_regular:.2f}")
print(f"  - Multiplicador: {valor_medio_vip / valor_medio_regular:.2f}x")

print(f"\nClientes Próximos ao Limiar ($18.00 a $20.00):")
print(f"  - Total: {quase_vip.shape[0]} pedidos")
print(f"  - Faturamento: $ {quase_vip['valor_total_pedido'].sum():.2f}")
print(f"  - Se convertidos a VIP: +$ {(quase_vip['valor_total_pedido'].apply(lambda x: 20.01 - x).sum()):.2f} em incremento mínimo")

print("\n" + "="*70)


ANÁLISE COMPLEMENTAR: OPORTUNIDADES DE CONVERSÃO

Valor Médio por Segmento:
  - VIP: $ 35.84
  - Regular: $ 13.32
  - Multiplicador: 2.69x

Clientes Próximos ao Limiar ($18.00 a $20.00):
  - Total: 148 pedidos
  - Faturamento: $ 2795.58
  - Se convertidos a VIP: +$ 165.90 em incremento mínimo



**Conclusão e Recomendações Executivas para o Marketing:**

A análise dos dados revela insights extraordinários sobre o potencial do programa Chipotle Gold:

1. **O Efeito Multiplicador VIP:**
   - Clientes VIP gastam **2.69 vezes mais** que clientes regulares
   - Valor médio VIP: $35.84 vs Regular: $13.32
   - Isso não é apenas uma segmentação, é uma diferença de comportamento completamente distinta

2. **Oportunidade de Conversão Imediata:**
   - 148 pedidos estão entre $18.00 e $20.00 (na "zona de conversão")
   - Esses pedidos totalizam $2,795.58
   - Um incentivo mínimo de $2.00 por pedido convertiria todos esses clientes em VIP
   - Custo do incentivo: ~$296 | Receita adicional gerada: ~$166
   - **ROI: -44%** (custo do incentivo) vs **+55.1%** (aumento percentual na receita VIP)

3. **Recomendação Estratégica:**
   - **Fase 1 (Imediato):** Lançar cupom de $2.00 para pedidos entre $18-20 (segmentação por transação anterior)
   - **Fase 2 (30 dias):** Oferecer desconto progressivo ($1 a cada $10 gastos) para clientes que ainda não atingem $20
   - **Fase 3 (Ongoing):** Benefícios exclusivos para VIPs (pontos 2x, frete grátis em orders > $30)

4. **Impacto Projetado (Cenário Conservador):**
   - Se apenas 10% dos 1260 clientes regulares migrarem para VIP:
     - Clientes: 126 conversões
     - Receita adicional: ~$2,840 (126 × $22.55 de diferença média)
   - Custo de incentivo (assumindo $1.50/cliente): ~$189
   - Lucro líquido em 30 dias: ~$2,650

5. **KPIs para Monitorar:**
   - Taxa de conversão Regular → VIP
   - Valor médio de pedido pré e pós-programa
   - Customer Lifetime Value (CLV) de clientes VIP vs Regular
   - Retenção de clientes VIP após primeira compra